agent/php_txn.c

```diff
@@ -536,6 +536,117 @@ static nrobj_t* nr_php_txn_get_labels() {
+
+  if (NULL == user_data) {
+    return NR_FAILURE;
+  }
+
+  nro_set_array_string(user_data, 0, key);
+
+  return NR_SUCCESS;
+}
+
+/*
+ * Purpose : Filter the labels hash to exclude any labels that are in the
+ *           newrelic.application_logging.forwarding.labels.exclude list.
+ *
+ * Params  : 1. The labels hash to filter.
+ *
+ * Returns : A new hash containing the filtered labels.
+ *           If no labels exist or all labels are excluded, then return NULL.
+ *
+ */
+
+static nrobj_t* nr_php_txn_get_log_forwarding_labels(nrobj_t* labels) {
+  nrobj_t* label_keys = NULL;
+  nrobj_t* exclude_labels_list = NULL;
+  nrobj_t* exclude_labels_hash = NULL;
+  nrobj_t* log_labels = NULL;
+
+  if (NULL == labels || 0 == nro_getsize(labels)) {
+    nrl_verbosedebug(NRL_TXN, "%s: No labels defined", __FUNCTION__);
+    return NULL;
+  }
+
+  /* if logging labels are disabled then nothing to do */
+  if (0 == NRINI(log_forwarding_labels_enabled)) {
+    nrl_verbosedebug(NRL_TXN, "%s: Log forwarding labels disabled",
+                     __FUNCTION__);
+    return NULL;
+  }
+
+  /* split exclude string on commas - nr_strsplit() will trim leading
+   * and trailing whitespace from each string extracted, as well as
+   * ignoring empty strings after whitespace trimming
+   */
+  exclude_labels_list
+      = nr_strsplit(NRINI(log_forwarding_labels_exclude), ",", 0);
+
+  /* convert to lowercase to support case insensitive search below
+   * will store lowercase version in a hash for more convenient lookup
+   */
+  exclude_labels_hash = nro_new(NR_OBJECT_HASH);
+  for (int i = 0; i < nro_getsize(exclude_labels_list); i++) {
+    char* label = nr_string_to_lowercase(
+        nro_get_array_string(exclude_labels_list, i + 1, NULL));
+
+    if (!nr_strempty(label)) {
+      nro_set_hash_boolean(exclude_labels_hash, label, 1);
+    }
+    nr_free(label);
+  }
+
+  /* original parsed exclude list is no longer needed */
+  nro_delete(exclude_labels_list);
+
+  /* collect label keys from existing labels */
+  label_keys = nro_new(NR_OBJECT_ARRAY);
+  nro_iteratehash(labels, nr_php_txn_collect_label_keys_iter,
+                  (void*)label_keys);
+
+  /* filter by going over the list of label keys, seeing if it exists in the
+   * exclude hash, and if it does skip it otherwise copy key/value for label
+   * to the log labels
+   */
+  log_labels = nro_new(NR_OBJECT_HASH);
+  for (int i = 0; i < nro_getsize(label_keys); i++) {
+    const char* key = nro_get_array_string(label_keys, i + 1, NULL);
+    const char* value = nro_get_hash_string(labels, key, NULL);
+    char* lower_key = nr_string_to_lowercase(key);
+    bool exclude = false;
+    nr_status_t rv;
+
+    exclude = nro_get_hash_boolean(exclude_labels_hash, lower_key, &rv);
+    if (NR_SUCCESS != rv) {
+      exclude = false;
+    }
+    if (!exclude) {
+      nro_set_hash_string(log_labels, key, value);
+    } else {
+      nrl_verbosedebug(NRL_TXN, "%s: Excluding label %s", __FUNCTION__, key);
```

Comment:
Key was not checked for NULL, this could generate a segfault.  Check for NULL earlier, or wrap in NRSAFESTR.

Reply:
I've added the NRSAFESTR macro just in case although with the change in https://github.com/newrelic/newrelic-php-agent/commit/c13d6c02f56cab1a6e3c0a011ab2966220771fde the value of `key` should not be `NULL` here.
See https://github.com/newrelic/newrelic-php-agent/pull/1027/commits/a238c676c9e40e1361284eec0f890d2b8a6fe0e6